contracts/AtomicSwapEther.sol

```diff
@@ -0,0 +1,108 @@
+pragma solidity ^0.5.2;
+
+contract AtomicSwapEther {
+    enum State {
+        NON_EXISTENT,
+        OPEN,
+        CLAIMED,
+        ABORTED
+    }
+
+    struct Swap {
+        State state;
+        address payable sender;
+        address payable recipient;
+        bytes32 hash;
+        uint256 amount;
+        uint256 timeout;
+        bytes32 preimage;
+    }
+
+    mapping (bytes32 => Swap) private swaps;
+
+    event Opened(bytes32 id, address recipient, bytes32 hash);
+    event Claimed(bytes32 id, bytes32 preimage);
+    event Aborted(bytes32 id);
+
+    modifier onlyExistentSwaps(bytes32 id) {
+        require(swaps[id].state != State.NON_EXISTENT, "No swap found for the given ID");
+        _;
+    }
+
+    modifier onlyNonExistentSwaps(bytes32 id) {
+        require(swaps[id].state == State.NON_EXISTENT, "Swap ID already exists");
+        _;
+    }
+
+    modifier onlyOpenSwaps(bytes32 id) {
+        require(swaps[id].state == State.OPEN, "No open swap found for the given ID");
+        _;
+    }
+
+    modifier onlyWithValidPreimage(bytes32 id, bytes32 preimage) {
+        require(swaps[id].hash == sha256(abi.encodePacked(preimage)), "Invalid preimage for swap hash");
+        _;
+    }
+
+    modifier onlyAbortableSwaps(bytes32 id) {
+        require(now >= swaps[id].timeout, "Swap timeout has not been reached");
+        _;
+    }
+
+    function open(
+        bytes32 id,
+        address payable recipient,
+        bytes32 hash,
+        uint256 timeout
+    ) external payable onlyNonExistentSwaps(id) {
+        swaps[id] = Swap({
+            state: State.OPEN,
+            sender: msg.sender,
+            recipient: recipient,
+            hash: hash,
+            amount: msg.value,
+            timeout: timeout,
+            preimage: bytes32(0)
+        });
+
+        emit Opened(id, recipient, hash);
+    }
+
+    function claim(
+        bytes32 id,
+        bytes32 preimage
+    ) external onlyOpenSwaps(id) onlyWithValidPreimage(id, preimage) {
+        swaps[id].state = State.CLAIMED;
+        swaps[id].preimage = preimage;
+
+        Swap memory swap = swaps[id];
+        swap.recipient.transfer(swap.amount);
+
+        emit Claimed(id, preimage);
+    }
+
+    function abort(
+        bytes32 id
+    ) external onlyOpenSwaps(id) onlyAbortableSwaps(id) {
+        swaps[id].state = State.ABORTED;
+
+        Swap memory swap = swaps[id];
+        swap.sender.transfer(swap.amount);
+
+        emit Aborted(id);
+    }
+
+    function viewSwap(
+        bytes32 id
+    ) external view onlyExistentSwaps(id) returns (
+        address sender,
+        address recipient,
+        bytes32 hash,
+        uint256 timeout,
+        uint256 amount,
+        bytes32 preimage
+    ) {
+        Swap memory swap = swaps[id];
```

Comment:
Did you find out why we cannot just write

```
return (
  swaps[id].sender,
  swaps[id].recipient,
  swaps[id].hash,
  swaps[id].timeout,
  swaps[id].amount,
  swaps[id].preimage,
)
```

Is the lookup expensive?

Reply:
I did this just so it was clear everything came from the same struct, but I'll look into the potential costs.

Reply:
I have no good reasons for one way or the other. Both look proper to me. I just remember we taked about an unnecessary copy here.